src/StreamJsonRpc/NerdbankMessagePackFormatter.cs

```diff
@@ -0,0 +1,2301 @@
+
+    /// <summary>
+    /// The constant "params", in its various forms.
+    /// </summary>
+    private static readonly CommonString ParamsPropertyName = new(Constants.Request.@params);
+
+    /// <summary>
+    /// The constant "traceparent", in its various forms.
+    /// </summary>
+    private static readonly CommonString TraceParentPropertyName = new(Constants.Request.traceparent);
+
+    /// <summary>
+    /// The constant "tracestate", in its various forms.
+    /// </summary>
+    private static readonly CommonString TraceStatePropertyName = new(Constants.Request.tracestate);
+
+    /// <summary>
+    /// The constant "2.0", in its various forms.
+    /// </summary>
+    private static readonly CommonString Version2 = new("2.0");
+
+    /// <summary>
+    /// A cache of property names to declared property types, indexed by their containing parameter object type.
+    /// </summary>
+    /// <remarks>
+    /// All access to this field should be while holding a lock on this member's value.
+    /// </remarks>
+    private static readonly Dictionary<Type, IReadOnlyDictionary<string, Type>> ParameterObjectPropertyTypes = new Dictionary<Type, IReadOnlyDictionary<string, Type>>();
+
+    /// <summary>
+    /// The serializer context to use for top-level RPC messages.
+    /// </summary>
+    private readonly FormatterContext rpcContext;
+
+    private readonly ProgressConverterResolver progressConverterResolver;
+
+    private readonly AsyncEnumerableConverterResolver asyncEnumerableConverterResolver;
+
+    private readonly PipeConverterResolver pipeConverterResolver;
+
+    private readonly MessagePackExceptionResolver exceptionResolver;
+
+    private readonly ToStringHelper serializationToStringHelper = new();
+
+    private readonly ToStringHelper deserializationToStringHelper = new();
+
+    /// <summary>
+    /// The serializer to use for user data (e.g. arguments, return values and errors).
+    /// </summary>
+    private FormatterContext userDataContext;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NerdbankMessagePackFormatter"/> class.
+    /// </summary>
+    public NerdbankMessagePackFormatter()
+    {
+        // Set up initial options for our own message types.
+        NBMP::MessagePackSerializer serializer = new()
+        {
+            InternStrings = true,
+            SerializeDefaultValues = false,
+        };
+
+        serializer.RegisterConverter(new RequestIdConverter());
+        serializer.RegisterConverter(new JsonRpcMessageConverter(this));
+        serializer.RegisterConverter(new JsonRpcRequestConverter(this));
+        serializer.RegisterConverter(new JsonRpcResultConverter(this));
+        serializer.RegisterConverter(new JsonRpcErrorConverter(this));
+        serializer.RegisterConverter(new JsonRpcErrorDetailConverter(this));
+        serializer.RegisterConverter(new TraceParentConverter());
```

Comment:
This requirement for state within each converter that is associated with each connection is going to cause extra startup cost for each connection/brokered service because the 'plan' for serialization will have to be recreated each time that the set of converters change.
I'm curious whether the `MessagePackSerializationContext` can help here, to provide some kind of reference to state associated with the connection that you control so that you don't have to recreate the converters each time.
To make a difference though, we'll need to not only be able to reuse converters but reuse the `MessagePackSerializer` object itself across connections. Do you think we could achieve that?

Tracking with https://github.com/AArnott/Nerdbank.MessagePack/issues/186

Reply:
I think the envelope serializer should be re-usable across connections. The set of envelope types is the same for all connections I think?

Reply:
Yes, that sounds like a good start.
Although the set of types to be serialized is probably larger on the user-defined side, so the startup hit will still be present.

Reply:
The StreamJsonRpc design is that every connection must have a unique instance of an IMessageFormatter. Given that, and that StreamJsonRpc has no insight into how callers may specially configure that formatter before adding it, I think your "FormatterContextBuilder" idea may be useful here. Suppose the StreamJsonRpc user create one FormatterContext which is then reused within each instance of NerdbankMessagePackFormatter. This reused context has a fully-configured MessagePackSerializer within it, so that that instance can be reused across connections.
So all you'd need at that point is a way to tunnel per-connection state into the custom converters you've registered within it (which themselves must then be stateless). Would that work?

Reply:
The per-connection state is the instance of NerdbankMessagePackFormatter?

I can expose the FormatterContextBuilder and FormatterContext as public types, and alter SetFormatterContext so that it takes a FormatterContext instance instead of a configuration lambda.

Would the per-connection state mostly take care of itself, since the FormatterContext would be set on each new instance of NerdbankMessagePackFormatter created by StreamJsonRpc?

Reply:
The per-connection state dependency is currently encapsulated in FormatterContextBuilder. We'll need to break that direct dependency, but I'm pretty confident we can do it.